File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<bits/stdc++.h>

using namespace std;

#define X first
#define Y second

int dx[4] = {1,-1,0,0};
int dy[4] = {0,0,1,-1};

queue<pair<int,int>> q;
int vis[105][105];

int solution(vector<vector<int>> maps)
{
    q.push({0,0});
    vis[0][0] = 1;
    
    int boardY = maps[0].size() - 1;
    
    while(!q.empty()) {
        auto t = q.front(); q.pop();
        int curX = t.X;
        int curY = t.Y;
        
        if(curX == maps.size()-1 && curY == boardY) {
            return vis[curX][curY];
        }
        
        for(int i=0; i<4; i++) {
            int nxtX = curX + dx[i];
            int nxtY = curY + dy[i];
            
            if(nxtX < 0 || nxtX >= maps.size() || nxtY < 0 || nxtY >= boardY+1) continue;
            if(vis[nxtX][nxtY] || !maps[nxtX][nxtY]) continue;
            
            q.push({nxtX,nxtY});
            vis[nxtX][nxtY] = vis[curX][curY] + 1;
        }
    }
    return -1;
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
import java.io.*;

class Solution {
    int[] dx = {1,-1,0,0};
    int[] dy = {0,0,1,-1};
    
    public int solution(int[][] maps) {
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{0,0,1});
        maps[0][0] = -1;
            
        while(!q.isEmpty()) {
            int[] cur = q.poll();
            int x = cur[0];
            int y = cur[1];
            int dis = cur[2];
            
            if(x == maps.length-1 && y == maps[0].length-1) {
                return dis;
            }
            
            for(int i=0; i<4; i++) {
                int nxtX = x + dx[i];
                int nxtY = y + dy[i];
                
                if(nxtX < 0 || nxtX >= maps.length || nxtY < 0 || nxtY >= maps[0].length) continue;
                if(maps[nxtX][nxtY] != 1) continue;
                
                maps[nxtX][nxtY] = dis+1;
                q.offer(new int[]{nxtX, nxtY, dis+1});
            }
        }
        return -1;
    }
}